### PR TITLE
implement get_link_whole_for() for IntelCompiler

### LINKED ISSUE #3862: intel compiler doesn't implement link_whole
Hi,
I got an error when including PETSc as external dependency. Here is my `meson.build`
```
project('xxx', 'cpp', default_options : ['cpp_std=c++11'])

pdep = dependency('PETSc')
```
this is what I get when I run meson

```The Meson build system
Version: 0.47.0
Source dir: /some/path/
Build dir: /some/path/build
Build type: native build
Project name: xxx
Project version: undefined
Native C++ compiler: icpc (intel 17.0.4 "icpc (ICC) 17.0.4 20170411")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.27.1)

meson.build:5:0: ERROR:  Language C++ does not support linking whole archives.

A full log can be found at /some/path/build/meson-logs/meson-log.txt
```
and the log is
```
Build started at 2018-07-10T11:48:00.454242
Main binary: /bin/python3
Python system: Linux
The Meson build system
Version: 0.47.0
Source dir: /some/path
Build dir: /some/path/build
Build type: native build
Project name: xxx
Project version: undefined
Sanity testing C++ compiler: icpc
Is cross compiler: False.
Sanity check compiler command line: icpc /some/path/build/meson-private/sanitycheckcpp.cc -o /some/path/build/meson-private/sanitycheckcpp.exe
Sanity check compile stdout:

-----
Sanity check compile stderr:

-----
Running test binary command: /some/path/build/meson-private/sanitycheckcpp.exe
Native C++ compiler: icpc (intel 17.0.4 "icpc (ICC) 17.0.4 20170411")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.27.1)
Determining dependency 'PETSc' with pkg-config executable '/usr/bin/pkg-config'
Called `/usr/bin/pkg-config --modversion PETSc` -> 0
3.9.2
Called `/usr/bin/pkg-config --cflags PETSc` -> 0
-I/...../petsc-3.9.2/arch-linux-c-dbg-g-real-avx-2-512-int64/include
Called `/usr/bin/pkg-config PETSc --libs` -> 0
-L/...../petsc-3.9.2/arch-linux-c-dbg-g-real-avx-2-512-int64/lib -lpetsc
Called `/usr/bin/pkg-config PETSc --libs` -> 0
-L/......./petsc-3.9.2/arch-linux-c-dbg-g-real-avx-2-512-int64/lib -lpetsc

meson.build:5:0: ERROR:  Language C++ does not support linking whole archives.
```

however, if I issue from the command line

```
icpc $(pkg-config --cflags --libs PETSc) meson-private/sanitycheckcpp.cc
```
it compiles and links..

What happens inside the call to `dependency`? Why I get that error and how can I solve it?

thanks

### PR DESCRIPTION
**Pull request by asartori86**
closes #3862

### CODE REVIEW

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -1561,6 +1561,9 @@ def openmp_flags(self):
         else:
             return ['-openmp']
 
+    def get_link_whole_for(self, args):
+        return ['-Wl,--whole-archive'] + args + ['-Wl,--no-whole-archive']
```
Comment (nirbheek) [inline mesonbuild/compilers/compilers.py]: I was actually thinking that this is a property of the linker, so the implementation will always be the same as the GnuCompiler. It makes sense to just use the GnuCompiler implementation in that case, like:

```
GnuCompiler.get_link_whole_for(self, args)
```

**Comment by nirbheek** [comment]
Another thing that would be a Very Nice Thing that ensures that stuff like this doesn't break the Intel compiler in the future is to find a way to add the latest version of ICC to the Docker instance that we use for running [Linux tests on the CI](https://github.com/mesonbuild/meson/blob/master/.travis.yml). Then we can run the tests against it in the same way that we run clang tests.

File: mesonbuild/compilers/compilers.py (hunk 1 above)
Comment (asartori86) [reply mesonbuild/compilers/compilers.py]: Ok I'll update

**Comment by asartori86** [comment]
> find a way to add the latest version of ICC to the Docker instance

don't you have to pay to have the intel compiler?

**Comment by nirbheek** [comment]
> don't you have to pay for intel compiler?

I believe you can get a free version for open source use: https://software.intel.com/en-us/qualify-for-free-software/opensourcecontributor

I remember getting it for FOSS use 1-2 years ago when I was implementing it for Meson, but I don't remember how.

**Comment by nirbheek** [comment]
The CI is a separate matter though, so if you have tested this patch and it works, we can merge it.

**Comment by asartori86** [comment]
Yes it worked for me

Il mar 10 lug 2018 21:48 Nirbheek Chauhan <notifications@github.com> ha
scritto:

**Pull request merged**